src/Coordination/KeeperDispatcher.cpp

```diff
@@ -51,6 +51,48 @@ namespace ErrorCodes
     extern const int SYSTEM_ERROR;
 }
 
+namespace
+{
+
+bool checkIfRequestIncreaseMem(const Coordination::ZooKeeperRequestPtr & request)
```

Comment:
But request to create a new session also increases memory usage https://github.com/ClickHouse/ClickHouse/blob/5a880cd0f019addce91f19e3dca6bbec72fb40f4/src/Coordination/KeeperDispatcher.cpp#L696-L749 and https://github.com/ClickHouse/ClickHouse/blob/5a880cd0f019addce91f19e3dca6bbec72fb40f4/src/Coordination/KeeperStorage.h#L142....

So here we will have situation:
1. We drop all connections with `Create|Multi` requests because of memory limit.
2. Client will try to reconnect and increase memory usage by `getSessionID` request.
3. Connection will be dropped again, loop here.

Also it worth to mention that read requests with watches also increase memory usage because we store watch inside internal map.

Reply:
Maybe add new error here https://github.com/ClickHouse/ClickHouse/blob/5a880cd0f019addce91f19e3dca6bbec72fb40f4/src/Common/ZooKeeper/IKeeper.h#L82-L91 like `ZOUTOFMEMORY` and use less active connection retries on clickhouse-server side?